### ISSUE TITLE
musl Haskell cross-build: <command line>: can't load .so/.DLL for: libgmp.so

### ISSUE THREAD

**Issue by nomeata** [NixOS/nixpkgs#57757]
## Issue description

I'd like to build Haskell programs statically, [e.g. `cachix`](https://github.com/cachix/cachix/issues/184) and others. I had good success using https://github.com/input-output-hk/haskell.nix/, but I’d like to be able to use plain nixpkgs as well.

So when trying to do a `musl64` cross build of Haskell libraries, some (but not all!) fail late during the build with the error message
```
<command line>: can't load .so/.DLL for: libgmp.so (libgmp.so: cannot open shared object file: No such file or directory)
```

This is after running haddock, so pretty late in the process. I can’t tell what exactly is trying to load `libgmp.so`.

### Steps to reproduce
```
nix-build  -A pkgsCross.musl64.haskellPackages.generics-sop
```
on current `master` (3aecf21239d105aadad229595348a0da4bcdbcd8)

## Technical details

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.0-2-amd64, Debian GNU/Linux, noversion`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.2.1`
 - nixpkgs: `/home/jojo/.nix-defexpr/channels/nixpkgs`
```
(that’s odd, I think I do have a multi-user setup. Certainly I need `NIX_REMOTE=daemon` to build anything)

**Comment by domenkozar** [comment]
cc @nh2 who has probably seen this one

**Comment by nh2** [comment]
@nomeata Have you seen https://github.com/nh2/static-haskell-nix ? That is designed to do exactly this.

In particular, it already provides static `cachix` builds; you can build it with a one-liner.

Your specific issue with `libgmp` is fixed [here](https://github.com/nh2/static-haskell-nix/blob/97dd9f518ba15b5b72d603cd9d029767a2b39227/survey/default.nix#L360) in there, one of many overrides for static libraries.

I plan to eventually make an overlay as part of the official `nixpkgs` on top of the new [`pkgsStatic`](https://github.com/NixOS/nixpkgs/pull/48803) work.

You can track the overall progress on #43795.

**Comment by nh2** [comment]
Oh and I shouldn't forget a plug:

I did most of this stuff on making static Haskell executables for a client, so just in case you need that for work and want commercial support / faster-than-free-time progress (e.g. for getting it into `nixpkgs` proper faster, or to port it to [`haskell.nix`](https://github.com/input-output-hk/haskell.nix/)), I can provide :)

Already [~95% of Stackage executables can be built this way](https://github.com/nh2/static-haskell-nix/issues/4#issuecomment-406838083), so I think we're on a good track.

**Comment by nomeata** [comment]
> @nomeata Have you seen https://github.com/nh2/static-haskell-nix ? That is designed to do exactly this.

I have! I got the impression that the support had already been merged into `nixpkgs`. I guess for making this official for `cachix` (which needs the binary in the main nixos cache) this is a prerequisite.

> Your specific issue with libgmp is fixed here in there, one of many overrides for static libraries.

Is that fix acceptable for mainline nixpkgs? Or would it have to be changed to only apply when building for musl?

> or to port it to haskell.nix

`haskell.nix` seems to already be in a good shape when it comes to static builds.

Sorry, so far I am not using this commercially :-)
